github/search.go

```diff
@@ -172,18 +179,52 @@ func (c CodeResult) String() string {
 // GitHub API docs: https://developer.github.com/v3/search/#search-code
 func (s *SearchService) Code(ctx context.Context, query string, opt *SearchOptions) (*CodeSearchResult, *Response, error) {
 	result := new(CodeSearchResult)
-	resp, err := s.search(ctx, "code", query, opt, result)
+	resp, err := s.search(ctx, "code", &searchParameters{Query: query}, opt, result)
+	return result, resp, err
+}
+
+// LabelsSearchResult represents the result of a code search.
+type LabelsSearchResult struct {
+	Total             *int          `json:"total_count,omitempty"`
+	IncompleteResults *bool         `json:"incomplete_results,omitempty"`
+	Labels            []LabelResult `json:"items,omitempty"`
+}
+
+// LabelResult represents a single search result.
+type LabelResult struct {
+	ID          *int64   `json:"id,omitempty"`
+	URL         *string  `json:"url,omitempty"`
+	Name        *string  `json:"name,omitempty"`
+	Color       *string  `json:"color,omitempty"`
+	Default     *bool    `json:"default,omitempty"`
+	Description *string  `json:"description,omitempty"`
+	Score       *float64 `json:"score,omitempty"`
+}
+
+func (l LabelResult) String() string {
+	return Stringify(l)
+}
+
+// Labels searches labels via various criteria.
+//
+// GitHub API docs: https://developer.github.com/v3/search/#search-labels
```

Comment:
We need a comment here about the significance of `repoID==0`.

Reply:
As I understand, `repoID` value of `0` is invalid. This endpoint requires a repository to be specified, and there is no repo with ID 0.

I haven't tried it, but I imagine that GitHub API would return an error.

We can still document it to make it more clear.

Reply:
Sorry I wasn't clear.  My comment was based on the code I saw below:

```go
	if parameters.RepositoryID != 0 {
		params.Set("repository_id", strconv.FormatInt(parameters.RepositoryID, 10))
	}
```

This shows that a `repoID` of `0` will not be sent at all.  That's all I meant by commenting it.
I was thinking of something like `// A repoID of 0 will not be sent.` or some such.

Also, @shurcooL - what are your thoughts about using `fmt.Sprintf` here instead of `strconv.FormatInt` ?
I prefer the former, but could go either way.

Reply:
Actually for that one I meant to have `repoID` As an optional parameter...
I guess that anything `<= 0` does not make sense for the API.

Maybe having that parameter as an `*int64` and checking for `nil` would make it clearer.

I have no opinion on `Sprintf` vs `FormatInt`. I’m d be curious to know what makes one better than the other.

Reply:
> I was thinking of something like `// A repoID of 0 will not be sent.` or some such.

A comment like that in `searchParameters` struct is okay with me.

However, it shouldn't be added to the `SearchService.Labels` endpoint because it's not a part of the API. `SearchService.Labels` always needs to be called with a valid `repoID`, and 0 is not a valid one.

An alternative to make things more clear could be to modify `searchParameters` to:

```Go
type searchParameters struct {
	Query        string
	RepositoryID *int64 // Sent if non-nil.
}
```

And `SearchService.Labels` would do:

```Go
&searchParameters{RepositoryID: &repoID, Query: query}
```

Reply:
I have turned `repoID` into an `*int64` I think it makes more sense since it’s an optional parameter in this API.

Reply:
I agree that it's optional in the context of `search` method (which is used for searching labels, repositories, issues, etc.), but is it optional for `Labels` method too?

From reading documentation at https://developer.github.com/v3/search/#search-labels, it sounded to me that it was mandatory to specify a repository when searching labels:

> Find labels in a repository with names or descriptions that match search keywords.

Have you tried if the endpoint works if repository ID is not specified?

Reply:
You are absolutely right. The doc does not explicitly states it is required like `q` for instance so I assumed it was not required.
I did get a 422 when querying without repo id. I’ll fix that in a minute.